src/main/java/org/folio/service/orders/CompositeOrderTotalFieldsPopulateService.java

```diff
@@ -88,33 +107,104 @@ private List<InvoiceLine> filterInvoiceLinesWithPoLines(List<InvoiceLine> invoic
 
   private Map<Invoice, List<InvoiceLine>> groupInvoiceLinesByInvoices(List<Invoice> invoices, List<InvoiceLine> invoiceLines) {
     return invoices.stream()
-      .collect(Collectors.toMap(invoice -> invoice, invoice -> invoiceLines.stream()
+      .collect(toMap(invoice -> invoice, invoice -> invoiceLines.stream()
         .filter(invoiceLine -> invoiceLine.getInvoiceId().equals(invoice.getId()))
         .collect(Collectors.toList())));
   }
 
-  private CompositeOrderRetrieveHolder populateTotalFields(CompositeOrderRetrieveHolder holder, Map<Invoice, List<InvoiceLine>> invoiceToInvoiceLinesMap, List<Transaction> transactions) {
+  private static class ExchangeRateMaps {
+    public Map<Invoice, ExchangeRate> invoiceRates;
+    public Map<Transaction, ExchangeRate> encumbranceRates;
+  }
+
+  private Future<ExchangeRateMaps> getExchangeRates(List<Invoice> invoices, List<Transaction> transactions,
+      RequestContext requestContext) {
+    return configurationEntriesCache.getSystemCurrency(requestContext)
+      .compose(systemCurrency -> getInvoiceExchangeRates(invoices, systemCurrency, requestContext)
+        .compose(invoiceRates -> getEncumbranceExchangeRates(transactions, systemCurrency,
+            requestContext)
+          .map(encumbranceRates -> {
+            ExchangeRateMaps maps = new ExchangeRateMaps();
+            maps.invoiceRates = invoiceRates;
+            maps.encumbranceRates = encumbranceRates;
+            return maps;
+          })
+      ));
+  }
+
+  private Future<Map<Invoice, ExchangeRate>> getInvoiceExchangeRates(List<Invoice> invoices, String systemCurrency,
```

Comment:
complexity is increased, what do you think about moving this new logic into new service?

Reply:
I could add these methods to `CacheableExchangeRateService`. Would that work for you ?

Reply:
okay, but if you see that during this refactoring these changes do not fit into CacheableExchangeRateService, you can remain as it, just wanted to highligth that this complex logic can be in some separate place, in this case it can be better tested with unit tests